sips/sip-022/sip-022-emergency-pox-fix.md

```diff
@@ -0,0 +1,235 @@
+
+* The total STX locked to the PoX address owned by caller of `stack-increase`
+  will be potentially higher than the amount of STX that the caller possesses.
+The caller will receive PoX payouts _as if_ they had locked that much STX.  So,
+the caller receives undue BTC.
+
+* The stacking threshold is raised to account for the PoX contract's reported
+  increase in STX locked.
+
+Furthermore, if this bug is triggered enough times, the Stacks network will crash.  This is
+because of a runtime assertion in the PoX reward set calculation logic that
+verifies that the total locked STX does not exceed the total liquid STX.  This
+would no longer be true due to this bug.  The offending assertion is detailed
+below:
+
+```rust
+pub fn get_reward_threshold_and_participation(
+    pox_settings: &PoxConstants,
+    addresses: &[RawRewardSetEntry],
+    liquid_ustx: u128,
+) -> (u128, u128) {
+    let participation = addresses
+        .iter()
+        .fold(0, |agg, entry| agg + entry.amount_stacked);
+
+    assert!(
+        participation <= liquid_ustx,
+        "CORRUPTION: More stacking participation than liquid STX"
+    );
+```
+
+The `RawRewardSetEntry` data are pulled directly from the
+`reward-cycle-pox-address-list` data map, and the `.amount_stacked` field is
+equal to the `total-ustx` field that was erroneously set in `stack-increase`.
+
+Considering these consequences with respect to the [blockchain catastrophic
+failure and recovery guidelines](https://github.com/stacksgov/sips/pull/10),
+this bug would be categorized as requiring an _immediate_ hard fork to rectify.
+THe network has not yet crashed, but it is in imminent danger of crashing and
+there is insufficient time to execute a fair, public SIP vote as has been customary for
+past hard forks. This SIP instead proposes that this hard fork activate at the
+start of the next reward cycle at the time of this writing.
+There is less than one reward cycle remaining to fix this
+problem, and yet a Stacker vote would require at least one complete reward cycle
+to carry out a STX coin vote (not accounting for any additional time required for
+sending out adequate public communications and tabulating the vote afterwards).
+
+# Specification
+
+Given the lack of time to conduct a fair, public vote to activate this SIP, the
+proposed fix in this SIP is as parsimonious and discreet as possible.  It will:
+
+* Disable `stack-increase` after the Bitcoin block activation height.  This
+  shall be achieved by altering the Clarity VM to declare that this function
+cannot be called, thereby preventing the buggy code body from being
+reached.  This shall not constitute a new version of the Clarity language;
+instead, the Stacks _chainstate specification_ shall be altered to require that
+this particular function in this particular contract after the Bitcoin block
+activation height is treated as absent from all Stacks forks.
+
+* Reset all `total-ustx` values in the `reward-cycle-pox-address-list` that
```

Comment:
Is this considered a change to the _chainstate specification_?
Something like `total-ustx` of `reward-cycle-pox-address-list` shall not be used to calculated the locked amount.

Reply:
Can you clarify in what way this change to the on-chain data is different to auto unlock?

Can we say that this SIP requires that the solo stacker lock amounts are auto unlocked earlier than usually and immediately relocked using a new calculation of the locking amount?

I would like to better understand why this SIP is different to arbitrary on-chain data manipulation (like set the balance of hacker A to 0). 

The current specification is too detailed/technical to understand it.

Reply:
Proposal for better understanding:

* Audit the locked amount for solo stackers and the corresponding amounts in the table `reward-cycle-pox-address-list` that is used to determine the number of reward slots. Correct the table entry if the locked amount of the users account (derived from using stack-stx, stack-increase and stack-extend transaction) do not match the stored total amount in the table `reward-cycle-pox-address-list`. The locked amount of the users' accounts shall be the correct value.  
  
  This correction can be seen as an auto-unlock at the time of activation followed by an immediate imaginary stack-stx transaction with the corrected amount. 
  
  This reset shall happen prior to any reward cycle processing in
cycle # 58 -- namely, prior to the call to
`get_reward_threshold_and_participation()` and prior to calculating any STX
auto-unlocks (which mutates state in `reward-cycle-pox-address-list`).
  
   The auditing and correction routine shall be invoked while
  processing every Stacks block whose parent Stacks block was mined in a Bitcoin
  block before the activation height.  Given that the set of solo stackers in
  cycle # 56 and cycle # 57 is small and fixed, the additional overhead of fixing all records in
  `reward-cycle-pox-address-list` will be minimal.  Nevertheless, as a safety
  precaution, each Stacks block whose parent was mined prior to the Bitcoin
  block activation height must contain zero transactions and confirm zero
  microblocks.

Reply:
This makes me think whether it is enough to unlock all solo stackers early and make them stack again for cycle 58